### ISSUE TITLE
Should an iterator with blank yield intent yield value tuples?

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#16233]
Currently, it looks like iterators with blank yield intent will yield mixed tuples where some elements are captured by reference and others by value. 

For example:

```chapel

record r { var x: int = 0; }

var g1 = new r();
var g2 = 0;

iter myIter1() {
  yield g1;
}

iter myIter2() {
  yield g2;
}

iter myIter3() {
  yield (g1, g2);
}

proc test1() {
  // Produces the error "cannot assign to const" variable for x.
  for x in myIter1() do x = new r(32);
  writeln(g1);
}

// Produces the error "cannot assign to const" variable for x.
proc test2() {
  for x in myIter2() do x = 32;
  writeln(g2);
}

proc test3() {
  for tup in myIter3() {
    tup[0] = new r(32);
    tup[1] = 32;
  }
  writeln(g1);  // Prints (x = 32)
  writeln(g2);  // Prints 0
}
test3();
```

In the example above, I create three iterators. The first yields a record by blank yield intent, the second an integer. In both cases the index variable is considered `const` and cannot be modified.

You can look at the AST to convince yourself, or use a second test to prove that the first two iterators yield new values and not references to existing ones.
<details>
  <summary>An example test </summary>

```chapel

record r { var x: int = 0; }

var g1 = new r();
var g2 = 0;

iter myIter1() {
  yield g1;
}

iter myIter2() {
  yield g2;
}

proc test1() {
  writeln("test1");
  for x in myIter1() {
    writeln(x);  // Prints (x = 0)
    g1 = new r(42);
    writeln(x);  // Prints (x = 0)
  }
}
test1();

proc test2() {
  writeln("test2");
  for x in myIter2() {
    writeln(x);  // Prints (0)
    g2 = 42;
    writeln(x);  // Prints (0)
  }
}
test2();
```
</details>

The third iterator yields tuples, and it is different from the first two iterators in two ways:

- It yields integer elements by value, and record elements by reference
- The index variable is not constant

My question is: **are either of these behaviors correct?**

## It yields integer elements by value, and record elements by reference

I propose that iterators with blank yield intent should yield value tuples in order to be consistent with iterators that yield other types by blank intent.

Much of the design rationale for tuples has envisioned them as functioning as a lightweight bundle of arguments. In which case, each element of a tuple yielded by blank intent should be captured as though it were being yielded by an iterator with the same type (for records and integers, blank intent yields new values).

In https://github.com/chapel-lang/chapel/issues/16231 I ask the question of whether or not iterator yield intents are the same as return intents. If they are, then I would expect the third iterator in this code to yield a value tuple, rather than a mixed tuple (a tuple where some elements are yielded by value, and some elements are returned by reference).

If we think the _current_ behavior is reasonable when yielding a tuple from an iterator with blank yield intent, we should go ahead and discuss it here or in #16231 in order to clarify the semantics in the spec.

## The index variable is not constant

I propose that tuple index variables yielded by blank, `const` or `const ref` yield intent should be constant just as they are for index variables of other types.

This may well just be a bug that occurs only when a mixed tuple is yielded by an iterator (a tuple that contains at least one reference seems not to be considered const).

**Comment by bradcray** [comment]
I'm not certain I agree that yielding or returning a tuple containing type `t` should behave the same as yielding or returning a value of type `t` directly.  Specifically, I think of a tuple's intention as being to bundle a group of related values together in a lightweight way, and arguably one that results in the least surprise.  If tucking an array into a tuple means doing a deep copy of that array, that's going to be expensive and potentially surprising, particularly if the array is large (say, a distributed 8TB array) whereas when tucking an integer into a tuple, the cheapest approach is presumably going to be to copy it in.  As far as I know, this choice of storing a copy vs. a reference in a tuple is currently consistent with our policy for how we pass variables of various types by default intent into a procedure or iterator.

That's just about creating the tuple though, whereas your question is almost as much about yielding (or presumably returning?) them.  In that case, I'm inclined to say a similar thing:  Even if I capture the 8TB array into the tuple by ref, if returning or yielding it is going to cause me to copy it out of the routine, that seems like a killer that I'd like to avoid by default.  So that suggests to me that that "default intent" interpretation should apply to each component of a tuple being returned/yielded, which suggests to me that tuple components of different types should be treated differently when yielded/returned by default yield/return intent.

I also wonder whether thinking about how tuples passed by `out` or `inout` intent might influence our interpretation of returning/yielding tuples.

**Comment by dlongnecke-cray** [comment]
> I'm not certain I agree that yielding or returning a tuple containing type t should behave the same as yielding or returning a value of type t directly. Specifically, I think of a tuple's intention as being to bundle a group of related values together in a lightweight way, and arguably one that results in the least surprise.

I think that "when yielding a tuple with blank intent, each individual element behaves as though it were yielded by an iterator directly" is pretty consistent with how we define intents for tuple formals.

> If tucking an array into a tuple means doing a deep copy of that array, that's going to be expensive and potentially surprising [...]

If I make an iterator yield an array today, it will return the array by value.

```chapel
var dom = {1..1};
var arr: [dom] int;

iter myIter() {
  yield arr;
}

proc test() {
  for a in myIter() {
    writeln(arr); // Prints 0
    a[1] = 64;
    writeln(arr); // Prints 0
  }
}
test();
```

So in that respect I think the surprising behavior is actually that tuples will return a reference to that array instead of a value.

> So that suggests to me that that "default intent" interpretation should apply to each component of a tuple being returned/yielded, which suggests to me that tuple components of different types should be treated differently when yielded/returned by default yield/return intent.

If you were to ask me "as an informed user" what I would expect the behavior for blank yield intent for tuples to be, I would base it off of how blank intent tuple arguments are designed, which is "apply the blank intent to each individual element". So by applying the blank yield intent to each element, I expect distinct values to be returned but actually get a mutable reference (where the mutability could be argued to just be a bug).

And if we want yield intents to be consistent with return intents, then the blank return intent for a tuple containing an array will also copy the array.

---

I do agree that it seems distasteful to copy out potentially large objects. My effort with relaxing the `ref` return intent for tuples is one non-optimal solution, but as we explored it might not be ideal for all situations or element types.

Another option in that case would be to explore a sort of individual element return intent, e.g. 

```chapel
iter foo() const: (int, ref int) {}
```

But that's a pretty big can of worms that I'm not sure we want to deal with this close to the release.

---

Or (as I suspect you are proposing here), we could go ahead and explore altering the blank return and yield intent for tuples in order to achieve a more "lightweight" behavior.

It _might_ be possible to come up with a set of rules for actual arguments to _build_tuple calls that enable them to be captured in an optimal manner, preserving refness for some and taking references to others, while converting other actuals to value.

E.g.

- If the actual is a call resulting in a `ref` then the reference will be preserved...
- If the actual is a global then a `ref` to it will be taken...

My worry here is w.r.t. to the constness of the index variable. In most scenarios a tuple index variable is const. What rules determine when it is const and when it is not? 

Another worry is preserving the constness of captured elements.

It seems like in order to be entirely correct, such a design would have to record the refness and the constness of each tuple element. So the "intent for individual elements" design, just implicitly instead of explicitly in the return type...

Or we could re-explore the `const` return intent and perhaps differentiate it from the blank return intent.

**Comment by vasslitvinov** [comment]
We have a couple of dimensions to be clear about:

(**Arg vs. Ret**) Passing an actual argument to a formal vs. returning, esp. by default intent. While we pass an 8TB array by reference, we return it by value. Ditto a record.
Yielding is similar to returning?

(**Lit vs. Var**) Is it a tuple literal `(a,b,c)` or a tuple variable `var tup = (a,b,c);` ?
A formal and an array element are also in the "a variable" category?
```chpl
return (myInt, myArray);
// vs.
var myTup = (myInt, myArray);
return myTup;
// how about refs?
ref myTupRef = (myInt, myArray);
return myTupRep;
```

> if I capture the 8TB array into the tuple by ref, if returning or yielding it is going to cause me to copy it out of the routine, that seems like a killer that I'd like to avoid by default. 

If I have a tuple with a ref in it and return/yield it by default intent, it is natural to expect that the ref will be (shallow-)copied, not the referenced thing. If my tuple contains an 8TB array, it is natural to expect that the array will be returned by value.

The only thing making me pause is:
```chpl
proc test() {
  ref r = myArray;
  return r;
}
```

where we return a copy of `myArray`, rather than a ref to it. Which is inconsistent with what I naturally expect from a tuple component.

**Comment by mppf** [comment]
I think changes in this area are likely to run into problems with the way that zippered iteration is implemented, but that is not a certainty.

It seems to me that `yield` should be similar to `return` in how the copies occur (if at all). Furthermore I would expect that somebody not wanting to copy an 8TB array would be careful to write code that yields/returns by `ref`. Or to use copy elision. Since yielding is different from returning, it is possible to have yielding of references to local variables in the iterator (say). However I don't think such things are a good idea (partly because we have parallel iterators and partly because I see more value in having simpler rules for when copies occur - making `return` and `yield` similar is simplifying the rules in my opinion).

We used to return arrays by `ref` by default but I changed that in PR #4762 as part of a campaign to make memory management for arrays be sane and not leaky. (For one thing, it does not make sense to me to have a different default ref-vs-value return intent for arrays than for regular records. I would not mind so much if it were only a question of constness).

We could consider having arrays return by `ref` intent again and rely on the lifetime checker to find errors. Here are some relevant examples for that:

``` chapel
proc returnsRefToArray() {
  var A:[1..1000] int;
  ref rA = A;
  return rA;
}
proc returnsRefToArray2() {
  var myObject = new ComplicatedObject();
  return myObject.getReferenceToArrayField();
}
```

Both of these functions would have to result in compilation error (from the lifetime checker). If they were allowed to compile and to return a reference, then they would return a reference to invalid memory.

Anyway, I think returning things by value by default is the right choice. (If we made the other choice, would we need to add a new return intent for "value" somehow?). (Another reason is that returning by value generally allows code to be written more easily but might require attention when optimizing. In contrast the above would be lifetime errors and the programmer would have to puzzle through why they do not work. I think we are making a language that allows easy prototyping followed by optimization - as opposed to say a language where the compiler only allows you to write things that it thinks are "fast").

So, to sum up, I think:
 * yield should behave similarly to return in terms of what copies are made
 * arrays should behave similarly to records in terms of what copies are made
 * behavior of an element for a returned tuple should be similar to behavior for returning the element by itself.

**Comment by bradcray** [comment]
> Even if I capture the 8TB array into the tuple by ref, if returning or yielding it is going to cause me to copy it out of the routine, that seems like a killer that I'd like to avoid by default.

I retract this statement.  You guys are correct that consistency with returning or yielding the array itself is more important than returning array tuple elements by ref by default.

**Comment by dlongnecke-cray** [comment]
It sounds like we're in agreement that an iterator with blank yield intent should yield value tuples?

**Comment by bradcray** [comment]
I'm not certain I know precisely what "value tuples" means, but if it equates to "the same thing that happens when procedures return tuples" then I'm on board.

**Comment by dlongnecke-cray** [comment]
The term "value tuples" is what we decided on in the spec to describe a tuple that contains no reference elements.

**Comment by dlongnecke-cray** [comment]
I think we agree that an iterator with the blank yield intent should yield value tuples. E.g. for the two definitions below...

```chapel
record r { var x: int = 0; }

iter foo() {
  yield (0, new r());
}

proc bar() {
  return (0, new r());
}
```

The type that is returned/yielded by both `foo` and `bar` should be the same.
